File: engine/src/main/java/com/terralite/engine/save/WorldSnapshot.java
package com.terralite.engine.save;

import com.terralite.engine.chunk.ChunkPos;

import java.util.List;
import java.util.Objects;

public record WorldSnapshot(List<ChunkPos> chunks, List<EntitySnapshot> entities, List<BlockSnapshot> blocks) {
    public WorldSnapshot(List<ChunkPos> chunks, List<EntitySnapshot> entities) {
        this(chunks, entities, List.of());
    }

    public WorldSnapshot {
        Objects.requireNonNull(chunks, "chunks");
        Objects.requireNonNull(entities, "entities");
        Objects.requireNonNull(blocks, "blocks");
        chunks = List.copyOf(chunks);
        entities = List.copyOf(entities);
        blocks = List.copyOf(blocks);
    }
}

File: engine/src/commonMain/kotlin/com/terralite/engine/save/WorldSnapshot.kt
package com.terralite.engine.save

import com.terralite.engine.chunk.ChunkPos

@JvmRecord
data class WorldSnapshot(
    val chunks: List<ChunkPos>,
    val entities: List<EntitySnapshot>,
    val blocks: List<BlockSnapshot>
) {
    constructor(chunks: List<ChunkPos>, entities: List<EntitySnapshot>) :
        this(chunks, entities, emptyList())
}
